### PR TITLE
Actor revalidation

### PR DESCRIPTION
**Pull request by pseusys**
# Description

Actor validation made required and also improved.

# Checklist

- [x] I have performed a self-review of the changes
- [x] Run lint
- [ ] Check all the tests + files for artifacts
- [x] Revise test actor (and maybe test pipeline)
- [x] Add docs  to functions

# To Consider

- Add tests (if functionality is changed)
- Update API reference / tutorials / guides
- Update CONTRIBUTING.md (if devel workflow is changed)
- Update `.ignore` files, scripts (such as `lint`), distribution manifest (if files are added/deleted)
- Search for references to changed entities in the codebase

### CODE REVIEW

**Review by ruthenian8 — CHANGES_REQUESTED** [review]

File: dff/pipeline/pipeline/actor.py [hunk 1]
```diff
@@ -56,6 +57,79 @@ def error_handler(error_msgs: list, msg: str, exception: Optional[Exception] = N
...
+
+def validate_callable(
+    callable: Callable,
+    name: str,
+    flow_label: str,
+    node_label: str,
+    error_msgs: list,
+    logging_flag: bool = True,
+    expected_types: Optional[List[Type]] = None,
+    return_type: Optional[Tuple[Type]] = None,
```
[the remark this answers is not in the document]
Comment (pseusys) [reply dff/pipeline/pipeline/actor.py:87]: Because `None` is also a type.
Otherwise, it would be impossible to distinguish cases when we:

1. Want to check return type, and the expected return type is `None`.
2. Do not want to check return type and thus passing `None` instead of the return type.

If we wrap return type into one-element `Tuple`, the difference becomes obvious: `None != (None,)`.

File: dff/pipeline/pipeline/actor.py [hunk 2]
```diff
@@ -325,68 +399,106 @@ def _choose_label(
...
+                        node.response,
+                        "response",
+                        flow_name,
+                        node_name,
+                        error_msgs,
+                        verbose,
+                        [Context, Pipeline, Any, Any],
+                        (Message,),
+                    )
+                elif node.response is not None and not issubclass(type(node.response), Message):
```
Comment (RLKRo) [inline dff/pipeline/pipeline/actor.py:450]:
```suggestion
                elif node.response is not None and not isinstance(node.response, Message):
```

File: dff/pipeline/pipeline/actor.py [hunk 3]
```diff
@@ -56,6 +57,79 @@ def error_handler(error_msgs: list, msg: str, exception: Optional[Exception] = N
         logger.error(msg, exc_info=exception)
 
 
+def types_match(type1: Type, type2: Type) -> bool:
+    """
+    This function compares types with assumption that one of the types might be a :py:class:`typing.ForwardRef`.
```
Comment (RLKRo) [inline dff/pipeline/pipeline/actor.py:62]: #310 removes forwardrefs.

File: dff/pipeline/pipeline/actor.py [hunk 4]
```diff
@@ -56,6 +57,79 @@ def error_handler(error_msgs: list, msg: str, exception: Optional[Exception] = N
...
+):
+    """
+    This function validates a function during :py:class:`~dff.script.Script` validation.
+    It checks parameter number (unconditionally), parameter types (if specified) and return type (if specified).
+
+    :param callable: Function to validate.
+    :param name: Type of the function (label, condition, response, etc.).
+    :param flow_label: Flow label this function is related to (used for error localization only).
+    :param node_label: Node label this function is related to (used for error localization only).
+    :param error_msgs: List that contains error messages. All the error message will be added to that list.
```
Comment (RLKRo) [inline dff/pipeline/pipeline/actor.py:97]: Why not return a list of error messages instead?

File: dff/pipeline/pipeline/actor.py [hunk 5]
```diff
@@ -56,6 +57,79 @@ def error_handler(error_msgs: list, msg: str, exception: Optional[Exception] = N
...
+    :param logging_flag: The flag which defines whether logging is necessary. Defaults to `True`.
+    :param expected_types: Tuple of types that correspond to expected function parameter types.
```
Comment (RLKRo) [inline dff/pipeline/pipeline/actor.py:99]: I think this function should infer `expected_types` and `return_type` from the `name` argument.
I don't like that expected function signatures are supposed to be passed from callers.

**Review by RLKRo — CHANGES_REQUESTED** [review]

File: dff/pipeline/pipeline/actor.py (hunk 4 above)
Comment (pseusys) [reply dff/pipeline/pipeline/actor.py:97]: It was designed for actor initially, the method returns boolean. Moreover, the error message list is used in the method.

File: dff/pipeline/pipeline/actor.py (hunk 5 above)
Comment (pseusys) [reply dff/pipeline/pipeline/actor.py:99]: Should they be extracted to some kind of external dictionary to keep expected argument types and return type?
The idea was to make this method as freestanding as possible, to  make it able to validate many different types of methods.

Comment (RLKRo) [reply dff/pipeline/pipeline/actor.py:99]: I think it would be better if mappings from function types to their signatures were defined in a single dictionary instead of being split across a hundred lines of code of `validate_script`.

If you would like for this method to remain freestanding, you can implement this with two functions -- the original `validate_callable` and a second one wrapping the first and acting as a proxy to fill out `expected_types` and `return_type` according to the dictionary of types.

File: dff/pipeline/pipeline/actor.py (hunk 4 above)
Comment (RLKRo) [reply dff/pipeline/pipeline/actor.py:97]: I meant that it is more pythonic for this function to return a list which is then appended to `error_msgs` in a caller function.

File: dff/pipeline/pipeline/actor.py [hunk 6]
```diff
@@ -60,6 +77,61 @@ def error_handler(error_msgs: list, msg: str, exception: Optional[Exception] = N
...
+    arguments_type, return_type = USER_FUNCTION_TYPES[function_type]
+    params = list(signature.parameters.values())
+    if len(params) != len(arguments_type):
+        msg = (
+            f"Incorrect parameter number of {name}={callable.__name__}: "
+            f"should be {len(arguments_type)}, found {len(params)}, "
+            f"error was found in (flow_label, node_label)={(flow_label, node_label)}"
+        )
+        error_handler(error_msgs, msg, None, logging_flag)
+    if TYPE_CHECKING:
```
Comment (RLKRo) [inline dff/pipeline/pipeline/actor.py:115]: This won't work. `TYPE_CHECKING` is always false at runtime.

**Review by RLKRo — CHANGES_REQUESTED** [review]
This version doesn't do annotation checks due to `TYPE_CHECKING` nature (it is always false).
I would suggest either importing the classes inside the validator function or comparing annotations using strings instead of types (something like `str(signature.parameters[0]) == "dff.script.core.Message"`).

Also, the validation checks should be moved from actor to script (directly into the `Node` class).
There will be a problem with validating labels (ensuring that constant labels point to an existing node).
`ContextVar` with `ValidationInfo` should solve this problem:
https://docs.pydantic.dev/latest/concepts/validators/#using-validation-context-with-basemodel-initialization
(save a list of all labels inside a before validator for the script model; access it from the label validation method of the `Node` class)

**Comment by RLKRo** [comment]
That's fair.
We can validate function signatures at `Node` level and transition labels at `Script` level.
(i.e. validators `validate_response_signature`, `validate_label_signature`, `validate_condition_signature`, `validate_processing_signatures` in `Node` and `check_labels_exist_in_script` for `Script`)

`Node` validation is easier to understand and debug: you define a function that checks a specific aspect of a node; and pydantic handles all the logic of validating it on its own -- we don't have to write our own loops.
This allows us to validate `Node` arguments even if they are not defined at the second level of the script. (which is not a common use case tbh)

**Comment by RLKRo** [comment]
@pseusys 
It doesn't seem like `test_script` is correct. There is indeed no "node" node from the `GLOBAL` node's perspective.
I assume that this test is supposed to test script creation.
Could you replace `test_script` (and remove the `script_test` function) with two tests (or one parametrized) that attempt to create a script object from `TOY_SCRIPT` and `MULTIFLOW_SCRIPT` objects of `dff.utils.testing.toy_script`?

File: dff/script/core/script.py [hunk 7]
```diff
@@ -25,6 +26,80 @@
...
+    :param expected_types: Tuple of types that correspond to expected function parameter types.
+        Also used for parameter number check. If `None`, parameter check is skipped. Defaults to `None`.
+    :param return_type: Tuple, containing expected function return type.
+        If `None` or contains more or less than 1 element, return type check is skipped. Defaults to `None`.
+    :return: list of produced error messages.
+    """
+    from dff.script.core.context import Context
+    from dff.pipeline.pipeline.pipeline import Pipeline
+
+    USER_FUNCTION_TYPES = {
```
Comment (RLKRo) [inline dff/script/core/script.py:58]: Define these as `Enum` and use pass `Enum` in this function instead of `str`.

File: dff/script/core/script.py [hunk 8]
```diff
@@ -25,6 +26,80 @@
...
+        "label": ((Context, Pipeline), None),
+        "response": ((Context, Pipeline), Message),
+        "condition": ((Context, Pipeline), bool),
+        "processing": ((Context, Pipeline), Context),
```
Comment (RLKRo) [inline dff/script/core/script.py:62]: Processing should return `None` since #253.

File: dff/script/core/script.py [hunk 9]
```diff
@@ -25,6 +26,80 @@
...
+    arguments_type, return_type = USER_FUNCTION_TYPES[function_type]
+    params = list(signature.parameters.values())
+    if len(params) != len(arguments_type):
+        msg = (
+            f"Incorrect parameter number of {name}={callable.__name__}: "
+            f"should be {len(arguments_type)}, found {len(params)}, "
+            f"error was found in (flow_label, node_label)={(flow_label, node_label)}"
+        )
+        error_handler(error_msgs, msg, None)
+    if TYPE_CHECKING:
```
Comment (RLKRo) [inline dff/script/core/script.py:77]: Again, this won't work due to `TYPE_CHECKING` always being false at runtime.
Please, add tests for every possible error message during script validation that we do indeed raise that error.

File: dff/script/core/script.py (hunk 7 above)
Comment (RLKRo) [reply dff/script/core/script.py:58]: Another option is to replace this line:
```
- function_type = name if name in USER_FUNCTION_TYPES.keys() else "processing"
+ if name not in USER_FUNCTION_TYPES:
+     raise ...
```
ATM if an incorrect type is passed we just ignore that.

File: dff/script/core/script.py (hunk 9 above)
Comment (RLKRo) [reply dff/script/core/script.py:77]: `TYPE_CHECKING` is not required in this function because you already import `Context` and `Pipeline` before defining `USER_FUNCTION_TYPES`. So there shouldn't be any problems if you simply remove the
`if TYPE_CHECKING` line in `validate_callable`.

**Comment by RLKRo** [comment]
Yep, seems good
Will likely approve soon (potentially with minor corrections).

**Comment by RLKRo** [comment]
Wait, no.
You forgot to add tests:
https://github.com/deeppavlov/dialog_flow_framework/pull/289#discussion_r1542659199

**Comment by RLKRo** [comment]
In case there was a misunderstanding:
I approve the current version of actor validation and would like to merge this PR once the tests are added.

File: tests/script/core/test_validation.py [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+def test_conditions():
+    param_number_script = {"root": { "start": { TRANSITIONS: { ("root", "start", 1): wrong_param_number } } } }
+    param_types_script = {"root": { "start": { TRANSITIONS: { ("root", "start", 1): wrong_param_types } } } }
+    return_type_script = {"root": { "start": { TRANSITIONS: { ("root", "start", 1): wrong_return_type } } } }
+    function_signature_test("wrong condition function signature", param_number_script, param_types_script, return_type_script)
+    try:
+        # wrong condition type
+        Script(script={"root": { "start": { TRANSITIONS: { ("root", "start", 1): 7 } } } })
+        raise Exception("can not be passed: wrong condition type")
+    except ValueError:
```
Comment (RLKRo) [inline tests/script/core/test_validation.py:92]: Don't spread bad testing practices.
Use [`pytest.raises`](https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest-raises).

File: tests/script/core/test_validation.py [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+def test_labels():
+    param_number_script = {"root": { "start": { TRANSITIONS: { wrong_param_number: exact_match(Message("hi")) } } } }
+    param_types_script = {"root": { "start": { TRANSITIONS: { wrong_param_types: exact_match(Message("hi")) } } } }
+    return_type_script = {"root": { "start": { TRANSITIONS: { wrong_return_type: exact_match(Message("hi")) } } } }
+    function_signature_test("wrong label function signature", param_number_script, param_types_script, return_type_script)
+    try:
+        # wrong label tuple flow
+        Script(script={"root": { "start": { TRANSITIONS: { ("other", "start", 1): exact_match(Message("hi")) } } } })
+        raise Exception("can not be passed: wrong label tuple flow")
+    except ValueError:
```
Comment (RLKRo) [inline tests/script/core/test_validation.py:58]: This code is unrelated to others in the same test suite.
Group multiple tests testing the same part into a class:
https://docs.pytest.org/en/7.1.x/getting-started.html#group-multiple-tests-in-a-class
Each method should test a separate case.
That way the cases become independent.

Comment (RLKRo) [reply tests/script/core/test_validation.py]: Yes.
Why wouldn't we want to use test classes? If the only reason is that we haven't used them before then we should start doing so.
They help with test case organization among other things.

> This code is actually related

What I meant is that these test cases are independent from each other (one is not a requirement for another).
Failing the first one would lead to us not even testing any of the other ones.

By creating a class `TestLabelValidation` we group the test cases in the intended pytest way -- they are still grouped but now they are executed separately.

**Review by RLKRo — APPROVED** [review]
Good :+1: 

Only some small suggestions left.

File: dff/script/core/script.py [hunk 12]
```diff
@@ -25,6 +27,97 @@
...
+    :param msg: Error message which is to be added into `error_msgs`.
+    :param exception: Invoked exception. If it has been set, it is used to obtain logging traceback.
+        Defaults to `None`.
+    :param logging_flag: The flag which defines whether logging is necessary. Defaults to `True`.
+    """
+    error_msgs.append(msg)
+    logger.error(msg, exc_info=exception)
+
+
+def _types_equal(signature_type: Any, expected_type: str) -> bool:
```
[the remark this answers is not in the document]
Comment (RLKRo) [reply dff/script/core/script.py:60]: There's a type alias for `NodeLabelXType`: `NodeLabelTupledType`.
One option is to insist on using that. I don't really like that -- unlike with `Message` we'd force user to import that type just for annotations.

Honestly, I feel like we shouldn't do return type validation for label functions at all. It feels either too restrictive (if we limit return types) or too error-prone (if we try to encompass most of them).

[the remark this answers is not in the document]
Comment (RLKRo) [reply dff/script/core/script.py:60]: OK.
But I don't like current type names.
How about we update `dff/script/core/types.py` with
```python
ConstLabel = Union[NodeLabelTupledType, str]
Label = Union[ConstLabel, Callable] # or Union[ConstLabel, Callable[[Context, Pipeline], ConstLabel] if possible
```
and also import them in `dff/script/__init__.py`.

Then `ConstLabel` would be the correct return type annotation for label functions validation.

[the remark this answers is not in the document]
Comment (RLKRo) [reply dff/script/core/script.py:47]: I can't find `Label` anywhere.

`LabelType` is not something we want to expose to users. We can rename it to `FlowIdentifier` and `NodeIdentifier` (so we'd have two types and script would become `Dict[FlowIdentifier, Dict[NodeIdentifier[Dict[Keywords, Any]]]`).
BTW we should rename `Keywords` to `Keyword`.

`NodeLabelType` already exists and is equal to what I propose to be named `Label`:
1. I don't like the `Type` suffix -- it shouldn't be there. We have `Message` and not `MessageType`.
2. I guess `NodeLabel` is fine.

I also think we should consider fundamentally renaming them (`NodeLabel`) to `TransitionAddress` or `Destination` / `TransitionDestination` in the future.

**Review by RLKRo — APPROVED** [review]
